src/dataflow/src/server.rs

```diff
@@ -522,41 +522,33 @@ where
         updates.clear();
     }
 
+    // TODO eli: this is where interesting frontier logging happens
     /// Send progress information to the coordinator.
     fn report_frontiers(&mut self) {
         if let Some(feedback_tx) = &mut self.feedback_tx {
             let mut upper = Antichain::new();
             let mut progress = Vec::new();
-            let ids = self
-                .render_state
-                .traces
-                .traces
-                .keys()
-                .cloned()
-                .collect::<Vec<_>>();
-            for id in ids {
-                if let Some(traces) = self.render_state.traces.get_mut(&id) {
-                    // Read the upper frontier and compare to what we've reported.
-                    traces.oks_mut().read_upper(&mut upper);
-                    let lower = self
-                        .reported_frontiers
-                        .get_mut(&id)
-                        .expect("Frontier missing!");
-                    if lower != &upper {
-                        let mut changes = ChangeBatch::new();
-                        for time in lower.elements().iter() {
-                            changes.update(time.clone(), -1);
-                        }
-                        for time in upper.elements().iter() {
-                            changes.update(time.clone(), 1);
-                        }
-                        let lower = self.reported_frontiers.get_mut(&id).unwrap();
-                        changes.compact();
-                        if !changes.is_empty() {
-                            progress.push((id, changes));
-                        }
-                        lower.clone_from(&upper);
+            for (id, traces) in self.render_state.traces.traces.iter_mut() {
```

Comment:
This change seems reasonable, but it was changed here https://github.com/MaterializeInc/materialize/commit/afe856a4d65d7141eeaa84022003fd254ea76f57#, and possibly intentionally. @frankmcsherry can you verify that there isn't some subtle behavior effected by cloning out the ids first? (I assume not, because the borrow checker generally protects against this kind of stuff, but you never know.)

Reply:
This seems pretty reasonable? I couldn't find the relevant change in the commit linked above, though I could imagine it was pretty obfuscated. But in any case, I don't believe the cloning of keys was important.